eth/watchers/senderwatcher.go

```diff
@@ -0,0 +1,248 @@
+func (sw *SenderWatcher) Watch() {
+	events := make(chan []*blockwatch.Event)
+	sub := sw.watcher.Subscribe(events)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-sw.quit:
+			return
+		case err := <-sub.Err():
+			log.Fatal(err)
+		case events := <-events:
+			sw.handleBlockEvents(events)
+		}
+	}
+}
+
+// Stop watching for events
+func (sw *SenderWatcher) Stop() {
+	close(sw.quit)
+}
+
+// Clear removes a key-value pair from the map
+// this can be invoked by the SenderMonitor cleanup loop
+func (sw *SenderWatcher) Clear(addr ethcommon.Address) {
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+	if _, ok := sw.senders[addr]; ok {
+		delete(sw.senders, addr)
+	}
+}
+
+func (sw *SenderWatcher) handleBlockEvents(events []*blockwatch.Event) {
+	for _, event := range events {
+		for _, log := range event.BlockHeader.Logs {
+			if event.Type == blockwatch.Removed {
+				log.Removed = true
+			}
+			if err := sw.handleLog(log); err != nil {
+				glog.Error(err)
+			}
+		}
+	}
+}
+
+func (sw *SenderWatcher) handleLog(log types.Log) error {
+	eventName, err := sw.dec.FindEventName(log)
+	if err != nil {
+		return fmt.Errorf("could not find event name: %v", err)
+	}
+
+	sw.mu.RLock()
+	defer sw.mu.RUnlock()
+	switch eventName {
+	case "DepositFunded":
+		var depositFunded contracts.TicketBrokerDepositFunded
+		if err := sw.dec.Decode("DepositFunded", log, &depositFunded); err != nil {
+			return fmt.Errorf("failed to decode DepositFunded event: %v", err)
+		}
+		sender := depositFunded.Sender
+		if sw.senders[sender] != nil {
+			sw.senders[sender].Deposit = sw.senders[sender].Deposit.Add(sw.senders[sender].Deposit, depositFunded.Amount)
+		} else if sw.senders[sender] == nil || log.Removed {
+			info, err := sw.lpEth.GetSenderInfo(sender)
+			if err != nil {
+				return fmt.Errorf("GetSenderInfo RPC call to remote node failed: %v", err)
+			}
+			sw.senders[sender] = info
+		}
+	case "ReserveFunded":
+		var reserveFunded contracts.TicketBrokerReserveFunded
+		if err := sw.dec.Decode("ReserveFunded", log, &reserveFunded); err != nil {
+			return fmt.Errorf("failed to decode ReserveFunded event: %v", err)
+		}
+		sender := reserveFunded.ReserveHolder
+		if sw.senders[sender] != nil {
+			sw.senders[sender].Reserve = sw.senders[sender].Reserve.Add(sw.senders[sender].Reserve, reserveFunded.Amount)
+		} else if sw.senders[sender] == nil || log.Removed {
+			info, err := sw.lpEth.GetSenderInfo(sender)
+			if err != nil {
+				return fmt.Errorf("GetSenderInfo RPC call to remote node failed: %v", err)
+			}
+			sw.senders[sender] = info
+		}
+	case "Withdrawal":
+		var withdrawal contracts.TicketBrokerWithdrawal
+		if err := sw.dec.Decode("Withdrawal", log, &withdrawal); err != nil {
+			return fmt.Errorf("failed to decode Withdrawal event: %v", err)
+		}
+		sender := withdrawal.Sender
+		if _, ok := sw.senders[sender]; ok && !log.Removed {
+			sw.senders[sender].Deposit = big.NewInt(0)
+			sw.senders[sender].Reserve = big.NewInt(0)
+		}
+		if log.Removed {
```

Comment:
We can do that but If we ever add additional events from `TicketBroker` to another watcher we will make RPC calls for removed logs even if they do not pertain to the `SenderWatcher` since the eventDecoder is bound to an ABI. 

Altough one way to circumvent this I think would be to add a default case that simply returns `nil` and doesn't do anything else ?

Reply:
I took a slightly different approach in bc0fff2 and interested to know what you think. 

I ended up extracting the logic for `log.Removed` out of the switch statement.

Now we will decode the event logs to get `sender` but we'll only parse state updates when `sender` already exists in the `SenderWatcher.senders` map and `log.Removed == false`.

Since we scope `sender` outside of the switch block now, we can handle `log.Removed == true` outside of the switch statement without duplicating the logic.

We can still extract this logic into a helper `fetchAndSetSenderInfo()` but it seems that would require some juggling with the locks and unlocks in `GetSenderInfo()` which might end up being uglier than just duplicating a few lines.